experimental/Schemes/IdealSheaves.jl

```diff
@@ -0,0 +1,470 @@
+end
+
+# this constructs the empty ideal sheaf
+function IdealSheaf(X::CoveredScheme) 
+  C = default_covering(X)
+  I = Dict{AbsSpec, Ideal}()
+  return IdealSheaf(X, C, I, check=false)
+end
+
+# internal routine to set up an ideal sheaf by automatic extension 
+# from one prescribed set of generators on one affine patch
+function IdealSheaf(X::CoveredScheme, C::Covering, U::Spec, g::Vector{RET}) where {RET<:MPolyElem}
+  C in coverings(X) || error("the covering does not belong to the scheme")
+  U in patches(C) || error("the affine open patch does not belong to the covering")
+  for f in g
+    parent(f) == ambient_ring(U) || error("the generators do not belong to the correct ring")
+  end
+  D = Dict{AbsSpec, Ideal}()
+  D[U] = ideal(OO(U), g)
+  D = extend!(C, D, check=false)
+  I = IdealSheaf(X, C, D, check=false)
+  return I
+end
+
+# more user facing routine with an automatic lookup of the associated affine patch.
+# Warning: This might be ambiguous and lead to unexpected results since one affine 
+# patch can be used in different coverings!
+function IdealSheaf(X::CoveredScheme, g::Vector{RET}) where {RET<:MPolyElem} 
+  length(g) == 0 && IdealSheaf(X)
+  R = parent(g[1])
+  for i in 2:length(g)
+    R == parent(g[i]) || error("elements do not belong to the same ring")
+  end
+  for C in coverings(X)
+    for U in patches(C)
+      R == ambient_ring(U) && return IdealSheaf(X, C, U, g)
+    end
+  end
+  error("the given set of generators could not be associated to an affine patch of the scheme")
+end
+
+function IdealSheaf(X::CoveredScheme, C::Covering, g::Vector{RET}) where {RET<:MPolyElem} 
+  length(g) == 0 && IdealSheaf(X)
+  R = parent(g[1])
+  for i in 2:length(g)
+    R == parent(g[i]) || error("elements do not belong to the same ring")
+  end
+  C in coverings(X) || error("covering is not listed")
+  for U in patches(C)
+    R == base_ring(OO(U)) && return IdealSheaf(X, C, U, g)
+  end
+  error("the given set of generators could not be associated to an affine patch of the scheme")
+end
+
+# pullback of an ideal sheaf for internal use between coverings of the same scheme
+function (F::CoveringMorphism)(I::IdealSheaf)
+  X = scheme(I)
+  D = codomain(F)
+  D == covering(I) || error("ideal sheaf is not defined on the correct covering")
+  C = domain(F)
+  new_dict = Dict{AbsSpec, Ideal}()
+
+  # go through the patches of C and pull back the generators 
+  # whenever they are defined on the target patch
+  for U in patches(C)
+    f = F[U]
+    V = codomain(f)
+    # for the basic patches here
+    if haskey(ideal_dict(I), V)
+      new_dict[U] = ideal(OO(U), pullback(f).(I[V]))
+    end
+    # check for affine refinements
+    if haskey(affine_refinements(D), V)
+      Vrefs = affine_refinements(D)[V]
+      # pull back the refinement
+      for W in Vrefs
+        h = pullback(f).(gens(W))
+        # take care to discard possibly empty preimages of patches
+        j = [i for i in 1:length(h) if !iszero(h)]
```

Comment:
please check if this works as intended. Looks curious that the `if` statement does not depend on `i`